vrrp: T6020: vrrp health-check script not applied correctly

HollyGurza opened the pull request:
T6020: vrrp health-check script not applied correctly in keepalived.conf

Added health check to sync-group in CLI
Don't use instance health-check when instance in sync group member
Warning notify for health-check config problem
New smoke test


<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->
Added health check to sync-group in CLI
Don't use instance health-check when instance in sync group member
Warning notify for health-check config problem
New smoke test
## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- All submitted PRs must be linked to a Task on Phabricator. -->
* https://vyos.dev/T6020

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->
## Proposed changes
<!--- Describe your changes in detail -->
high-availability vrrp

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->

## Smoketest result
<!-- Provide the output of the smoketest
```
$ /usr/libexec/vyos/tests/smoke/cli/test_xxx_feature.py
test_01_simple_options (__main__.TestFeature.test_01_simple_options) ... ok
```
-->
```
vyos@vyos# python3 /usr/libexec/vyos/tests/smoke/cli/test_high-availability_vrrp.py 
test_01_default_values (__main__.TestVRRP.test_01_default_values) ... ok
test_02_simple_options (__main__.TestVRRP.test_02_simple_options) ... ok
test_03_sync_group (__main__.TestVRRP.test_03_sync_group) ... ok
test_04_exclude_vrrp_interface (__main__.TestVRRP.test_04_exclude_vrrp_interface) ... ok
test_check_health_script (__main__.TestVRRP.test_check_health_script) ... ok

----------------------------------------------------------------------
Ran 5 tests in 43.278s

OK

```

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly

HollyGurza committed:
vrrp: T6020: vrrp health-check script not applied correctly in keepalived.conf

Added health-check to sync-group in CLI
Don't use instance health-check when instance in sync group member
Disallow wrong healtch-check configurations
New smoke test

zdc reviewed (changes requested):
In the history of this template, I see that this logic was changed several times, many of them with mistakes.

Would it not be better to finally add health-check to sync-group in CLI and disallow configurations with health-check inside the vrrp group if it is a group member?

c-po commented on data/templates/high-availability/keepalived.conf.j2:
```diff
@@ -128,7 +128,7 @@ vrrp_instance {{ name }} {
     }
 {%         endif %}
-{%         if group_config.health_check is vyos_defined %}
+{%         if group_config.health_check is vyos_defined and group_config.health_check.script is not vyos_defined %}
```
`{%         if group_config.health_check.script is not vyos_defined %}` would be sufficient in this case

faekz0r commented:
> In the history of this template, I see that this logic was changed several times, many of them with mistakes.
> 
> Would it not be better to finally add health-check to sync-group in CLI and disallow configurations with health-check inside the vrrp group if it is a group member?

That sounds like a great idea. In fact when I ran into the current bug, I was astounded that you can't actually configure the health-check for the vrrp group from the VyOS CLI.
Otherwise I could have also worked around the bug by configuring it correctly from CLI.

HollyGurza commented:
> In the history of this template, I see that this logic was changed several times, many of them with mistakes.
> 
> Would it not be better to finally add health-check to sync-group in CLI and disallow configurations with health-check inside the vrrp group if it is a group member?

Done

sever-sever commented:
@zdc Could you check the proposed changes?

dmbaturin reviewed (approved):
I don't see any issues with the implementation.
@zdc could you check if everything you wanted to address is done?

HollyGurza committed:
vrrp: T6020: vrrp health-check script not applied correctly in keepalived.conf

Added health-check to sync-group in CLI
Don't use instance health-check when instance in sync group member
Disallow wrong healtch-check configurations
New smoke test

dmbaturin commented:
@mergify backport sagitta

dmbaturin commented:
@mergifyio backport sagitta

sever-sever commented:
@MergifyIo backport sagitta

sever-sever commented:
@HollyGurza Could you create PR for 1.4? 
For some reason, mergify does not work

HollyGurza commented:
> @HollyGurza Could you create PR for 1.4? For some reason, mergify does not work

https://github.com/vyos/vyos-1x/pull/3122
